platforms/core-runtime/launcher/src/main/java/org/gradle/tooling/internal/provider/ProviderConnection.java

```diff
@@ -346,10 +347,28 @@ private Parameters initParams(ProviderOperationParameters operationParameters) {
             daemonParams.setBaseDir(operationParameters.getDaemonBaseDir());
         }
 
-        //override the params with the explicit settings provided by the tooling api
-        List<String> jvmArguments = operationParameters.getJvmArguments();
-        if (jvmArguments != null) {
-            daemonParams.setJvmArgs(jvmArguments);
+        // Since 8.13, we split the daemon JVM args into base and additional JVM args (#31462)
+        // When calling the following new methods, the following states can happen:
+        // - incompatible (before 8.13 provider): falling back to old behavior,
+        // - not used: no methods should be called, default daemon JVM arguments are applied
+        // - used:
+        //   - base JVM arguments (if supplied) override the default daemon JVM arguments
+        //   - additional JVM (if supplied) applied on top of the set JVM arguments
+        try {
+            List<String> baseJvmArguments = operationParameters.getBaseJvmArguments();
+            if (baseJvmArguments != null && !baseJvmArguments.isEmpty()) {
+                daemonParams.setJvmArgs(baseJvmArguments);
+            }
+            List<String> additionalJvmArguments = operationParameters.getAdditionalJvmArguments();
+            if (additionalJvmArguments != null && !additionalJvmArguments.isEmpty()) {
+                daemonParams.addJvmArgs(additionalJvmArguments);
+            }
```

Comment:
This is correct, but the `isEmpty()` check is probably redundant as `addJvmArgs(emptyList())` is, iiuc, a no-op.